ampd/src/event_sub/event_stream.rs

```diff
@@ -0,0 +1,272 @@
+use std::iter;
+use std::time::Duration;
+
+use axelar_wasm_std::FnExt;
+use error_stack::ResultExt;
+use events::Event;
+use futures::{stream, Stream, StreamExt};
+use tendermint::block;
+use tokio::time::{interval, Interval};
+
+use crate::tm_client::TmClient;
+
+type Error = super::Error;
+type Result<T> = error_stack::Result<T, Error>;
+
+pub async fn event_stream<'a, T, F>(
+    tm_client: &'a T,
+    poll_interval: Duration,
+    capacity: usize,
+    skip_block_fn: F,
+) -> Result<impl Stream<Item = Result<Event>> + 'a>
+where
+    T: TmClient,
+    F: Fn() -> bool + Copy + 'a,
+{
+    Ok(block_stream(tm_client, poll_interval)
+        .await?
+        .map(move |block_height| process_block(tm_client, block_height, skip_block_fn))
+        .buffered(capacity)
+        .flat_map(|result| {
+            result.map_or_else(
+                |err| stream::iter(vec![Err(err)]),
+                |events| stream::iter(events.into_iter().map(Ok).collect::<Vec<_>>()),
+            )
+        }))
+}
+
+async fn process_block<T>(
+    tm_client: &T,
+    block_height: Result<block::Height>,
+    skip_block_fn: impl Fn() -> bool,
+) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    match (skip_block_fn(), block_height) {
+        (true, _) => Ok(vec![]),
+        (_, Ok(block_height)) => events(tm_client, block_height).await,
+        (_, Err(err)) => Err(err),
+    }
+}
+
+async fn block_stream<'a, T>(
+    tm_client: &'a T,
+    poll_interval: Duration,
+) -> Result<impl Stream<Item = Result<block::Height>> + 'a>
+where
+    T: TmClient,
+{
+    latest_block_height(tm_client)
+        .await
+        .map(|latest| BlockState {
+            current: latest,
+            latest,
+        })
+        .map(|block_state| block_state.stream(tm_client, interval(poll_interval)))
```

Comment:
I'm not sure this works as you expect it to. This will create a new interval for each block height, instead of reusing the same one

Reply:
This `.map` is applied on a `Result` not a stream, so I'm not sure how does it "create a new interval for each block height"?